# Section 1.1: A deep(er) dive into joining and cleaning datasets

As a general rule, data scientists assume that about 80 percent of the time and effort devoted to a project will be spent preparing data for anlaysis. Because most real-world data is split among multiple existing datasets, this invariably means cleaning and joining different datasets together in the way that a data scientist needs for analysis. Thus, mastering these skills is essential for undertaking data science.

This section makes extensive use of pandas, the principal Python library for data handling and manipulation. Note that this section assumes some familiarity with pandas and basic Python skills.

In order to provide an experience more like real-world data science, we will use real data taken gathered from the [U.S. Department of Agriculture National Nutrient Database for Standard Reference](https://www.ars.usda.gov/northeast-area/beltsville-md-bhnrc/beltsville-human-nutrition-research-center/nutrient-data-laboratory/docs/usda-national-nutrient-database-for-standard-reference/).

## Reminders about importing, built-in Help, and documentation

The standard convention in Python-centric data science is to import pandas under the alias `pd`, which is what we will use here:

In [1]:
import pandas as pd

Because this is such a common convention, it is the way we will use and refer to pandas throughout the rest of this section and this course. You should also adopt this usage in your own code to make it easily readable for other data scientists.

Pandas is a big package and there can be a lot to keep track of. Fortunately, IPython (the underlying program that powers this notebook and other like it) gives you the ability to quickly explore the contents of a package like pandas by using its tab-completion feature. If you want to see all of the functions available with pandas, type this:

```ipython
In [2]: pd.<TAB>

```

When you do so, a drop-down menu will appear next to the `pd`.

> **Exercise**

In [2]:
# Place your cursor after the period and press <TAB>:
pd.concat?

# Now select an item using tab-completion and then add a period
# and use tab-completion to explore *that*.
# For example, you could try placing pressing <TAB> after:
# pd.DataFrame.


As we progress through Section 1.1, don't forget that IPython also provides a tab-completion feature and function documentation with the `?` (question mark) character. If you don't understand something about a function you see in this section, taking a moment to csonult the documentation can help a great deal. You will find this documentation to be a very valuable reference source for your own data science work, both now and in the future. As a reminder, use this code to display the built-in pandas documentation:

```ipython
In [4]: pd?
```

> **Exercise**

> Run this code cell and review the documentation for the pandas DataFrame object. We are going to use it quite a bit.

In [3]:
pd?

## A brief reminder about Jupyter notebooks

This course makes extensive use of Jupyter notebooks hosted on Microsoft Azure. Azure-hosted Jupyter notebooks provide an easy way for you to experiment with programming concepts in an interactive fashion that requires no installation of software by students on local computers.

Jupyter notebooks are divided into cells. Each cell either contains text written in the Markdown markup language or a space in which to write and execute computer code. Because all the code resides inside code cells, you can run each code cell inline rather than using a separate Python interactive window.

> **Note**: This notebook is designed to have you run code cells one by one, and several code cells contain deliberate errors for demonstration purposes. As a result, if you use the **Cell** > **Run All** command, some code cells past the error won't be run. To resume running the code in each case, use **Cell** > **Run All Below** from the cell after the error.

## Loading data

> **Learning goal:** By the end of this subsection, you should be comfortable loading data from files into panda `DataFrame`s and troubleshooting any difficulties that might arise.

Because pandas `DataFrame`s are two-dimensional data structures, they are inherently similar to flat-file formats such as comma-separated value (CSV) files, the most common import and export format for spreadsheets and databases. Adding to this ease of translation from CSV files to `DataFrame`s, pandas provides a convenient function to load the contents of CSV files into `DataFrame`s (more convenient, in fact, then the native Python [CSV library](https://docs.python.org/3.6/library/csv.html)). Let's get comfortable with [pandas.read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) because we will be using often.

In [4]:
df = pd.read_csv('Data/USDA-nndb.csv')

## Exploring and cleaning the data

> **Learning goal:** By the end of this subsection, you should be comfortable performing simple exploration of your data and performing simple cleaning steps on it to prepare it for later analysis.

Let's now take a look at our `DataFrame`:

In [5]:
df.head()

NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0    1001        BUTTER,WITH SALT      15.87         717         0.85   
1    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
2    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
3    1004             CHEESE,BLUE      42.41         353        21.40   
4    1005            CHEESE,BRICK      41.11         371        23.24   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
0          81.11     2.11            0.06           0.0           0.06   
1          78.30     1.62            2.87           0.0           0.06   
2          99.48     0.00            0.00           0.0           0.00   
3          28.74     5.11            2.34           0.0           0.50   
4          29.68     3.18            2.79           0.0           0.51   

      ...      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
0     ...             7.0      51.368       21.021        3.043   
1     ...             4.6      45.390       19.874        3.331   
2     ...             8.6      61.924       28.732        3.694   
3     ...             2.4      18.669        7.778        0.800   
4     ...             2.5      18.764        8.598        0.784   

   Cholestrl_(mg)  GmWt_1                  GmWt_Desc1  GmWt_2  \
0           215.0    5.00  1 pat,  (1" sq, 1/3" high)    14.2   
1           225.0    3.80  1 pat,  (1" sq, 1/3" high)     9.4   
2           256.0   12.80                      1 tbsp   205.0   
3            75.0   28.35                        1 oz    17.0   
4            94.0  132.00                1 cup, diced   113.0   

        GmWt_Desc2  Refuse_Pct  
0           1 tbsp         0.0  
1           1 tbsp         0.0  
2            1 cup         0.0  
3     1 cubic inch         0.0  
4  1 cup, shredded         0.0  

[5 rows x 53 columns]

We can get some more specific information about the `DataFrame` by using its `info()` method:

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8790 entries, 0 to 8789
Data columns (total 53 columns):
NDB_No               8790 non-null int64
Shrt_Desc            8790 non-null object
Water_(g)            8789 non-null float64
Energ_Kcal           8790 non-null int64
Protein_(g)          8790 non-null float64
Lipid_Tot_(g)        8790 non-null float64
Ash_(g)              8465 non-null float64
Carbohydrt_(g)       8790 non-null float64
Fiber_TD_(g)         8196 non-null float64
Sugar_Tot_(g)        6958 non-null float64
Calcium_(mg)         8442 non-null float64
Iron_(mg)            8646 non-null float64
Magnesium_(mg)       8051 non-null float64
Phosphorus_(mg)      8211 non-null float64
Potassium_(mg)       8364 non-null float64
Sodium_(mg)          8707 non-null float64
Zinc_(mg)            8084 non-null float64
Copper_mg)           7533 non-null float64
Manganese_(mg)       6630 non-null float64
Selenium_(�g)        7090 non-null float64
Vit_C_(mg)           7972 non-null flo

Just quickly inspecting the columns from `df` we can see that almost all of the columns have a number of null values. Those missing values are not an issue for us right now, but they will pose a challenge in future sections (but we will deal with them in those sections).

Let's also check to see if this `DataFrame` has any duplicate values in it. We'll do this by looking at the `NDB_No` column, which serves as the unique identifier for all foods in the National Nutrient Database for Standard Reference.

In [7]:
df.duplicated('NDB_No').sum()

0

Given the nature of the data source (a government reference database) it makes sense that there are no duplicate entries. For purposes of learning more about cleaning data, let's duplicate data by using the `append()` method.

In [8]:
df = df.append(df, ignore_index=True)

The `append()` method has basically stacked the `DataFrame` by appending a copy of `df` to the end of the `DataFrame`. The `ignore_index=True` parameter means that the index numbering for the newly doubled `DataFrame` continues seamlessly.

Now let's look directly at how many times individual values in a column (such as `NDB_No`) are duplicated using the `pivot_table` method for the `df` `DataFrame`:

In [9]:
df = df.drop_duplicates('NDB_No', keep="last")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8790 entries, 8790 to 17579
Data columns (total 53 columns):
NDB_No               8790 non-null int64
Shrt_Desc            8790 non-null object
Water_(g)            8789 non-null float64
Energ_Kcal           8790 non-null int64
Protein_(g)          8790 non-null float64
Lipid_Tot_(g)        8790 non-null float64
Ash_(g)              8465 non-null float64
Carbohydrt_(g)       8790 non-null float64
Fiber_TD_(g)         8196 non-null float64
Sugar_Tot_(g)        6958 non-null float64
Calcium_(mg)         8442 non-null float64
Iron_(mg)            8646 non-null float64
Magnesium_(mg)       8051 non-null float64
Phosphorus_(mg)      8211 non-null float64
Potassium_(mg)       8364 non-null float64
Sodium_(mg)          8707 non-null float64
Zinc_(mg)            8084 non-null float64
Copper_mg)           7533 non-null float64
Manganese_(mg)       6630 non-null float64
Selenium_(�g)        7090 non-null float64
Vit_C_(mg)           7972 non-null

The `DataFrame` is now half of its original size, which is what we would expect. However, look at this line in the `df.info()` output:

`Int64Index: 8790 entries, 8790 to 17579`

While there are only now 8790 entries per column, the indexing for the DataFrame does not run 0 through 8789, as we might have expected. We can see this more directly by looking at the `head` of the redacted `DataFrame`:

In [10]:
df.head()

NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
8790    1001        BUTTER,WITH SALT      15.87         717         0.85   
8791    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
8792    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
8793    1004             CHEESE,BLUE      42.41         353        21.40   
8794    1005            CHEESE,BRICK      41.11         371        23.24   

      Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
8790          81.11     2.11            0.06           0.0           0.06   
8791          78.30     1.62            2.87           0.0           0.06   
8792          99.48     0.00            0.00           0.0           0.00   
8793          28.74     5.11            2.34           0.0           0.50   
8794          29.68     3.18            2.79           0.0           0.51   

         ...      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
8790     ...             7.0      51.368       21.021        3.043   
8791     ...             4.6      45.390       19.874        3.331   
8792     ...             8.6      61.924       28.732        3.694   
8793     ...             2.4      18.669        7.778        0.800   
8794     ...             2.5      18.764        8.598        0.784   

      Cholestrl_(mg)  GmWt_1                  GmWt_Desc1  GmWt_2  \
8790           215.0    5.00  1 pat,  (1" sq, 1/3" high)    14.2   
8791           225.0    3.80  1 pat,  (1" sq, 1/3" high)     9.4   
8792           256.0   12.80                      1 tbsp   205.0   
8793            75.0   28.35                        1 oz    17.0   
8794            94.0  132.00                1 cup, diced   113.0   

           GmWt_Desc2  Refuse_Pct  
8790           1 tbsp         0.0  
8791           1 tbsp         0.0  
8792            1 cup         0.0  
8793     1 cubic inch         0.0  
8794  1 cup, shredded         0.0  

[5 rows x 53 columns]

> **Takeaway:** Duplicate, `Null`, and `NaN` values can all complicate (if not derail) your analysis. Learning how to identify and remove these problems is a huge part of successfully performing data science.

## Splitting the `DataFrame`

> **Learning goal:** By the end of this subsection, you should be comfortable selecting and dropping specific columns from a `DataFrame`.

It might seem strange to discuss splitting a `DataFrame` in a course section on joining them, but we'll do so here to create the `DataFrame`s that we'll join later on. We take this approach for two reasons:

1. Creating our own `DataFrame`s gives us easy control over the content of the child `DataFrame`s to best demonstrate aspects of joining datasets.
2. Because we have a baseline, joined `DataFrame` (`df`), it's easy to see how different methods of joining the child `DataFrame`s produce different results.

We'll create two child `DataFrame`s from `df`: `df1` and `df2`. First, we'll create `df1` from the first 35 columns of `df`.

In [11]:
df1 = df.iloc[:,:35]

> **Exercise**

> Why did we use syntax `df1 = df.iloc[:,:35]` to capture the first 35 columns of `df`? What does the first `:` (colon) in the square brackets do? Experiment with `df3 = df.iloc[:35]` in the code cell below and compare `df3.info()` with `df1.info()` to satisfy yourself as to why we need to use this syntax.

In [ ]:
#listcompresion

In [12]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8790 entries, 8790 to 17579
Data columns (total 35 columns):
NDB_No               8790 non-null int64
Shrt_Desc            8790 non-null object
Water_(g)            8789 non-null float64
Energ_Kcal           8790 non-null int64
Protein_(g)          8790 non-null float64
Lipid_Tot_(g)        8790 non-null float64
Ash_(g)              8465 non-null float64
Carbohydrt_(g)       8790 non-null float64
Fiber_TD_(g)         8196 non-null float64
Sugar_Tot_(g)        6958 non-null float64
Calcium_(mg)         8442 non-null float64
Iron_(mg)            8646 non-null float64
Magnesium_(mg)       8051 non-null float64
Phosphorus_(mg)      8211 non-null float64
Potassium_(mg)       8364 non-null float64
Sodium_(mg)          8707 non-null float64
Zinc_(mg)            8084 non-null float64
Copper_mg)           7533 non-null float64
Manganese_(mg)       6630 non-null float64
Selenium_(�g)        7090 non-null float64
Vit_C_(mg)           7972 non-null

We'll create `df2` in a similar manner to `df1`, but we need to do things a little differently here to ensure that the first columne (`NDB_No`) makes it into `df2`. This is going to serve as the column that's common to both child `DataFrame`s when we join them later in this section.

We also want to populate `df2` with a different number of rows than `df1`. Doing so will make is easier to demonstrate what goes on with some of the join techniques shown below.

In [13]:
df2 = df.iloc[0:2000, [0]+[i for i in range(35,53)]]

> **Question**

> If you're unsure about why we use `range(35,53)` in the list comprehension above, review the documentation for the `range()` function using `range?` in the code cell below.

In [14]:
# Remeber that Python uses zero-indexing.


We can examine `df2` by using the `head()` and `info()` methods.

In [15]:
df2.head()

NDB_No  Alpha_Carot_(�g)  Beta_Carot_(�g)  Beta_Crypt_(�g)  \
8790    1001               0.0            158.0              0.0   
8791    1002               1.0            135.0              6.0   
8792    1003               0.0            193.0              0.0   
8793    1004               0.0             74.0              0.0   
8794    1005               0.0             76.0              0.0   

      Lycopene_(�g)  Lut+Zea_ (�g)  Vit_E_(mg)  Vit_D_�g  Vit_D_IU  \
8790            0.0            0.0        2.32       0.0       0.0   
8791            0.0           13.0        1.37       0.0       0.0   
8792            0.0            0.0        2.80       0.0       0.0   
8793            0.0            0.0        0.25       0.5      21.0   
8794            0.0            0.0        0.26       0.5      22.0   

      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  Cholestrl_(mg)  \
8790         7.0      51.368       21.021        3.043           215.0   
8791         4.6      45.390       19.874        3.331           225.0   
8792         8.6      61.924       28.732        3.694           256.0   
8793         2.4      18.669        7.778        0.800            75.0   
8794         2.5      18.764        8.598        0.784            94.0   

      GmWt_1                  GmWt_Desc1  GmWt_2       GmWt_Desc2  Refuse_Pct  
8790    5.00  1 pat,  (1" sq, 1/3" high)    14.2           1 tbsp         0.0  
8791    3.80  1 pat,  (1" sq, 1/3" high)     9.4           1 tbsp         0.0  
8792   12.80                      1 tbsp   205.0            1 cup         0.0  
8793   28.35                        1 oz    17.0     1 cubic inch         0.0  
8794  132.00                1 cup, diced   113.0  1 cup, shredded         0.0

In [16]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2000 entries, 8790 to 10789
Data columns (total 19 columns):
NDB_No              2000 non-null int64
Alpha_Carot_(�g)    1426 non-null float64
Beta_Carot_(�g)     1447 non-null float64
Beta_Crypt_(�g)     1425 non-null float64
Lycopene_(�g)       1426 non-null float64
Lut+Zea_ (�g)       1418 non-null float64
Vit_E_(mg)          1456 non-null float64
Vit_D_�g            1289 non-null float64
Vit_D_IU            1289 non-null float64
Vit_K_(�g)          1359 non-null float64
FA_Sat_(g)          1965 non-null float64
FA_Mono_(g)         1823 non-null float64
FA_Poly_(g)         1825 non-null float64
Cholestrl_(mg)      1947 non-null float64
GmWt_1              1967 non-null float64
GmWt_Desc1          1967 non-null object
GmWt_2              1171 non-null float64
GmWt_Desc2          1171 non-null object
Refuse_Pct          2000 non-null float64
dtypes: float64(16), int64(1), object(2)
memory usage: 312.5+ KB


Let’s take a look at `df1`.

In [17]:
df1.head()

NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
8790    1001        BUTTER,WITH SALT      15.87         717         0.85   
8791    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
8792    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
8793    1004             CHEESE,BLUE      42.41         353        21.40   
8794    1005            CHEESE,BRICK      41.11         371        23.24   

      Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
8790          81.11     2.11            0.06           0.0           0.06   
8791          78.30     1.62            2.87           0.0           0.06   
8792          99.48     0.00            0.00           0.0           0.00   
8793          28.74     5.11            2.34           0.0           0.50   
8794          29.68     3.18            2.79           0.0           0.51   

          ...       Vit_B6_(mg)  Folate_Tot_(�g)  Folic_Acid_(�g)  \
8790      ...             0.003              3.0              0.0   
8791      ...             0.008              4.0              0.0   
8792      ...             0.001              0.0              0.0   
8793      ...             0.166             36.0              0.0   
8794      ...             0.065             20.0              0.0   

      Food_Folate_(�g)  Folate_DFE_(�g)  Choline_Tot_ (mg)  Vit_B12_(�g)  \
8790               3.0              3.0               18.8          0.17   
8791               4.0              4.0               18.8          0.07   
8792               0.0              0.0               22.3          0.01   
8793              36.0             36.0               15.4          1.22   
8794              20.0             20.0               15.4          1.26   

      Vit_A_IU  Vit_A_RAE  Retinol_(�g)  
8790    2499.0      684.0         671.0  
8791    2468.0      683.0         671.0  
8792    3069.0      840.0         824.0  
8793     721.0      198.0         192.0  
8794    1080.0      292.0         286.0  

[5 rows x 35 columns]

You'll notice on that both `DataFrame`s have their old indices indexes that they inherited from  `df`. We can fix that by using the `reset_index()` method, but then we run into a problem.

In [18]:
df1 = df1.reset_index()
df1.head()

index  NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0   8790    1001        BUTTER,WITH SALT      15.87         717         0.85   
1   8791    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
2   8792    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
3   8793    1004             CHEESE,BLUE      42.41         353        21.40   
4   8794    1005            CHEESE,BRICK      41.11         371        23.24   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)      ...       \
0          81.11     2.11            0.06           0.0      ...        
1          78.30     1.62            2.87           0.0      ...        
2          99.48     0.00            0.00           0.0      ...        
3          28.74     5.11            2.34           0.0      ...        
4          29.68     3.18            2.79           0.0      ...        

   Vit_B6_(mg)  Folate_Tot_(�g)  Folic_Acid_(�g)  Food_Folate_(�g)  \
0        0.003              3.0              0.0               3.0   
1        0.008              4.0              0.0               4.0   
2        0.001              0.0              0.0               0.0   
3        0.166             36.0              0.0              36.0   
4        0.065             20.0              0.0              20.0   

   Folate_DFE_(�g)  Choline_Tot_ (mg)  Vit_B12_(�g)  Vit_A_IU  Vit_A_RAE  \
0              3.0               18.8          0.17    2499.0      684.0   
1              4.0               18.8          0.07    2468.0      683.0   
2              0.0               22.3          0.01    3069.0      840.0   
3             36.0               15.4          1.22     721.0      198.0   
4             20.0               15.4          1.26    1080.0      292.0   

   Retinol_(�g)  
0         671.0  
1         671.0  
2         824.0  
3         192.0  
4         286.0  

[5 rows x 36 columns]

Our old indexes are still there for `df1`, but now they're in a new column titled "`index`." We need to address that by using the `drop=True` parameter for the method. (We also need to drop the `index` column we just created.)

In [19]:
df1 = df1.drop(['index'], axis=1)
df1 = df1.reset_index(drop=True)
df1.head()

NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0    1001        BUTTER,WITH SALT      15.87         717         0.85   
1    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
2    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
3    1004             CHEESE,BLUE      42.41         353        21.40   
4    1005            CHEESE,BRICK      41.11         371        23.24   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
0          81.11     2.11            0.06           0.0           0.06   
1          78.30     1.62            2.87           0.0           0.06   
2          99.48     0.00            0.00           0.0           0.00   
3          28.74     5.11            2.34           0.0           0.50   
4          29.68     3.18            2.79           0.0           0.51   

       ...       Vit_B6_(mg)  Folate_Tot_(�g)  Folic_Acid_(�g)  \
0      ...             0.003              3.0              0.0   
1      ...             0.008              4.0              0.0   
2      ...             0.001              0.0              0.0   
3      ...             0.166             36.0              0.0   
4      ...             0.065             20.0              0.0   

   Food_Folate_(�g)  Folate_DFE_(�g)  Choline_Tot_ (mg)  Vit_B12_(�g)  \
0               3.0              3.0               18.8          0.17   
1               4.0              4.0               18.8          0.07   
2               0.0              0.0               22.3          0.01   
3              36.0             36.0               15.4          1.22   
4              20.0             20.0               15.4          1.26   

   Vit_A_IU  Vit_A_RAE  Retinol_(�g)  
0    2499.0      684.0         671.0  
1    2468.0      683.0         671.0  
2    3069.0      840.0         824.0  
3     721.0      198.0         192.0  
4    1080.0      292.0         286.0  

[5 rows x 35 columns]

Now let's do the same thing to `df2`.

In [21]:
df2 = df2.reset_index(drop=True)
df2.head()

NDB_No  Alpha_Carot_(�g)  Beta_Carot_(�g)  Beta_Crypt_(�g)  Lycopene_(�g)  \
0    1001               0.0            158.0              0.0            0.0   
1    1002               1.0            135.0              6.0            0.0   
2    1003               0.0            193.0              0.0            0.0   
3    1004               0.0             74.0              0.0            0.0   
4    1005               0.0             76.0              0.0            0.0   

   Lut+Zea_ (�g)  Vit_E_(mg)  Vit_D_�g  Vit_D_IU  Vit_K_(�g)  FA_Sat_(g)  \
0            0.0        2.32       0.0       0.0         7.0      51.368   
1           13.0        1.37       0.0       0.0         4.6      45.390   
2            0.0        2.80       0.0       0.0         8.6      61.924   
3            0.0        0.25       0.5      21.0         2.4      18.669   
4            0.0        0.26       0.5      22.0         2.5      18.764   

   FA_Mono_(g)  FA_Poly_(g)  Cholestrl_(mg)  GmWt_1  \
0       21.021        3.043           215.0    5.00   
1       19.874        3.331           225.0    3.80   
2       28.732        3.694           256.0   12.80   
3        7.778        0.800            75.0   28.35   
4        8.598        0.784            94.0  132.00   

                   GmWt_Desc1  GmWt_2       GmWt_Desc2  Refuse_Pct  
0  1 pat,  (1" sq, 1/3" high)    14.2           1 tbsp         0.0  
1  1 pat,  (1" sq, 1/3" high)     9.4           1 tbsp         0.0  
2                      1 tbsp   205.0            1 cup         0.0  
3                        1 oz    17.0     1 cubic inch         0.0  
4                1 cup, diced   113.0  1 cup, shredded         0.0

For practice, let's export these `DataFrame`s to CSV files by using the `to_csv()` method. Note that unless we explicitly tell pandas not to, it will also export the index as a column in the CSV file. We will also need to be careful to explicitly encode our CSV to UTF-8.

In [22]:
df1.to_csv('Data/NNDB1.csv', sep=',', encoding='utf-8',index=False)

> **Exercise**

In [23]:
# Export df2 to a CSV file.


> **Takeaway:** Although it's not common in the real world to split `DataFrame`s only to re-merge them later, you'll need to drop columns or create new `DataFrame`s that contain only the information you need. With truly large datasets, this is not just a convenience for you analysis, but a necessity for memory and performance!

## Joining `DataFrame`s

> **Learning goal:** By the end of this subsection, you should be comfortable performing left, right, inner, and outer merges on `DataFrame`s.

We'll examine the  most commonly used `DataFrame` function for joining datasets: `merge()`. But first, let's refresh ourselves on the shapes of our two `DataFrame`s so that the output of our joining makes more sense.

In [24]:
df1.shape

(8790, 35)

In [ ]:
df2.shape

The type of dataset join that’s most widely used by practicing data scientists is the left join. If you already have some experience with SQL, you know what this refers to. Basically, a left join is a join that takes all of the data from one `DataFrame` (think of it as the left set in a Venn diagram) and merges it with everything that it has in common with another `DataFrame` (the intersection with the right set in the same Venn diagram).

We do this using the `merge()` function. We also need to specify the type of join we want to perform by using the `how` parameter, as well as the index on which to join the `DataFrames` by using the `on` parameter.

In [25]:
left_df = pd.merge(df1, df2, how='left', on='NDB_No')
left_df.shape

(8790, 53)

> **Question**

> Is the shape of the resulting `DataFrame` what you were expecting? Why or why not?

Now let's compare this to the original `df` `DataFrame`.

In [26]:
df.shape

(8790, 53)

> **Question**

> The shapes are the same, but do you expect `df` and `left_df` to be identical? If so, why? If not, what differences do you expect there to be between them?

Let's check to see what the differences between these `DataFrame`s might be.

In [27]:
df.head()

NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
8790    1001        BUTTER,WITH SALT      15.87         717         0.85   
8791    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
8792    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
8793    1004             CHEESE,BLUE      42.41         353        21.40   
8794    1005            CHEESE,BRICK      41.11         371        23.24   

      Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
8790          81.11     2.11            0.06           0.0           0.06   
8791          78.30     1.62            2.87           0.0           0.06   
8792          99.48     0.00            0.00           0.0           0.00   
8793          28.74     5.11            2.34           0.0           0.50   
8794          29.68     3.18            2.79           0.0           0.51   

         ...      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
8790     ...             7.0      51.368       21.021        3.043   
8791     ...             4.6      45.390       19.874        3.331   
8792     ...             8.6      61.924       28.732        3.694   
8793     ...             2.4      18.669        7.778        0.800   
8794     ...             2.5      18.764        8.598        0.784   

      Cholestrl_(mg)  GmWt_1                  GmWt_Desc1  GmWt_2  \
8790           215.0    5.00  1 pat,  (1" sq, 1/3" high)    14.2   
8791           225.0    3.80  1 pat,  (1" sq, 1/3" high)     9.4   
8792           256.0   12.80                      1 tbsp   205.0   
8793            75.0   28.35                        1 oz    17.0   
8794            94.0  132.00                1 cup, diced   113.0   

           GmWt_Desc2  Refuse_Pct  
8790           1 tbsp         0.0  
8791           1 tbsp         0.0  
8792            1 cup         0.0  
8793     1 cubic inch         0.0  
8794  1 cup, shredded         0.0  

[5 rows x 53 columns]

In [28]:
left_df.head()

NDB_No               Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0    1001        BUTTER,WITH SALT      15.87         717         0.85   
1    1002  BUTTER,WHIPPED,W/ SALT      16.72         718         0.49   
2    1003    BUTTER OIL,ANHYDROUS       0.24         876         0.28   
3    1004             CHEESE,BLUE      42.41         353        21.40   
4    1005            CHEESE,BRICK      41.11         371        23.24   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
0          81.11     2.11            0.06           0.0           0.06   
1          78.30     1.62            2.87           0.0           0.06   
2          99.48     0.00            0.00           0.0           0.00   
3          28.74     5.11            2.34           0.0           0.50   
4          29.68     3.18            2.79           0.0           0.51   

      ...      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
0     ...             7.0      51.368       21.021        3.043   
1     ...             4.6      45.390       19.874        3.331   
2     ...             8.6      61.924       28.732        3.694   
3     ...             2.4      18.669        7.778        0.800   
4     ...             2.5      18.764        8.598        0.784   

   Cholestrl_(mg)  GmWt_1                  GmWt_Desc1  GmWt_2  \
0           215.0    5.00  1 pat,  (1" sq, 1/3" high)    14.2   
1           225.0    3.80  1 pat,  (1" sq, 1/3" high)     9.4   
2           256.0   12.80                      1 tbsp   205.0   
3            75.0   28.35                        1 oz    17.0   
4            94.0  132.00                1 cup, diced   113.0   

        GmWt_Desc2  Refuse_Pct  
0           1 tbsp         0.0  
1           1 tbsp         0.0  
2            1 cup         0.0  
3     1 cubic inch         0.0  
4  1 cup, shredded         0.0  

[5 rows x 53 columns]

The indexes notwithstanding, the first five rows of both `DataFrame`s are the same. Let's check the last five rows.

In [29]:
df.tail()

NDB_No                   Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
17575   83110             MACKEREL,SALTED      43.00         305        18.50   
17576   90240  SCALLOP,(BAY&SEA),CKD,STMD      70.25         111        20.54   
17577   90480                  SYRUP,CANE      26.00         269         0.00   
17578   90560                   SNAIL,RAW      79.20          90        16.10   
17579   93600            TURTLE,GREEN,RAW      78.50          89        19.80   

       Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
17575          25.10    13.40            0.00           0.0            0.0   
17576           0.84     2.97            5.41           0.0            0.0   
17577           0.00     0.86           73.14           0.0           73.2   
17578           1.40     1.30            2.00           0.0            0.0   
17579           0.50     1.20            0.00           0.0            0.0   

          ...      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
17575     ...             7.8       7.148        8.320        6.210   
17576     ...             0.0       0.218        0.082        0.222   
17577     ...             0.0       0.000        0.000        0.000   
17578     ...             0.1       0.361        0.259        0.252   
17579     ...             0.1       0.127        0.088        0.170   

       Cholestrl_(mg)  GmWt_1                          GmWt_Desc1  GmWt_2  \
17575            95.0    80.0  1 piece,  (5-1/2" x 1-1/2" x 1/2")    17.0   
17576            41.0    85.0                                3 oz     NaN   
17577             0.0    21.0                           1 serving     NaN   
17578            50.0    85.0                                3 oz     NaN   
17579            50.0    85.0                                3 oz     NaN   

                   GmWt_Desc2  Refuse_Pct  
17575  1 cubic inch, boneless         0.0  
17576                     NaN         0.0  
17577                     NaN         0.0  
17578                     NaN         0.0  
17579                     NaN         0.0  

[5 rows x 53 columns]

In [30]:
left_df.tail()

NDB_No                   Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
8785   83110             MACKEREL,SALTED      43.00         305        18.50   
8786   90240  SCALLOP,(BAY&SEA),CKD,STMD      70.25         111        20.54   
8787   90480                  SYRUP,CANE      26.00         269         0.00   
8788   90560                   SNAIL,RAW      79.20          90        16.10   
8789   93600            TURTLE,GREEN,RAW      78.50          89        19.80   

      Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
8785          25.10    13.40            0.00           0.0            0.0   
8786           0.84     2.97            5.41           0.0            0.0   
8787           0.00     0.86           73.14           0.0           73.2   
8788           1.40     1.30            2.00           0.0            0.0   
8789           0.50     1.20            0.00           0.0            0.0   

         ...      Vit_K_(�g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
8785     ...             NaN         NaN          NaN          NaN   
8786     ...             NaN         NaN          NaN          NaN   
8787     ...             NaN         NaN          NaN          NaN   
8788     ...             NaN         NaN          NaN          NaN   
8789     ...             NaN         NaN          NaN          NaN   

      Cholestrl_(mg)  GmWt_1  GmWt_Desc1  GmWt_2  GmWt_Desc2  Refuse_Pct  
8785             NaN     NaN         NaN     NaN         NaN         NaN  
8786             NaN     NaN         NaN     NaN         NaN         NaN  
8787             NaN     NaN         NaN     NaN         NaN         NaN  
8788             NaN     NaN         NaN     NaN         NaN         NaN  
8789             NaN     NaN         NaN     NaN         NaN         NaN  

[5 rows x 53 columns]

There are differences here in the last five rows. Notice that the right-most columns of `left_df` contain have Not a Number (`NaN`) values. This is because the left `DataFrame` was larger than the right `DataFrame`.

> **Exercise**

> A right join is simply the mirror image of a left join in which those entries from the left `DataFrame` that are common with the right `DataFrame` are merged with the right `DataFrame`.
>
>Perform a right join of `df1` and `df2` in the code cell below. But before you do that, ask yourself what shape you expect the resulting `DataFrame` to have? Do you expect it to have any `NaN` values?

In [31]:
# Hint: the parameter for the right join is how='right'


Another intuitive and widely used type of join is the inner join. This join simply merges entries that are common to both `DataFrame`s, resulting in a `DataFrame` that has no `NaN` values.

In [32]:
inner_df = pd.merge(df1, df2, how='inner', on='NDB_No')

> **Question**

> Before we examine the shape of the resulting `DataFrame`, what do you predict it will be? Why?

In [33]:
inner_df.shape

(2000, 53)

Did `inner_df` behave as you expected it would? Let's briefly examine it by using the `head()` and `tail()` methods.

In [ ]:
inner_df.head()

In [ ]:
inner_df.tail()

The resulting `DataFrame` is essentially the first 2000 rows of the original `df` `DataFrame`.

> **Exercise**

> An outer join is the union of two `DataFrame`s; anything that is in either `DataFrame` with be in the resultant one. Perform an outer join of `df1` and `df2`. What shape do you expect the resulting `DataFrame` to have? How does it differ from the right join of `df1` and `df2`? What differences would there have to be in the shape or content of either `DataFrame` for the outer join of the two to be different from their right join?

## Preparing for coming sections

We will be using the USDA NNDB dataset in Sections 1.2 and 1.3. However, particularly in Section 1.2, we want to include food group information to go with the food entries to aid with interpreting the result of our data analysis in that section. You will add food group information to this USDA dataset in preparation for these coming sections.

First, let's reload our original NNDB dataset so that we have a clean copy.

In [ ]:
df = pd.read_csv('Data/USDA-nndb.csv', encoding='latin_1')

Now let's load in the columns that we want from the older NNDB dataset that includes food groups.

In [ ]:
fg_df = pd.read_csv('Data/USDA-nndb-combined.csv', usecols=['NDB_No', 'FoodGroup'])
fg_df.info()
#fg_df = fg_df.rename(index=str, columns={' ID': 'NDB_No'})
#fg_df.head()

Note that `fg_df` does not have the same number of rows as `df`:

In [ ]:
fg_df.shape

> **Exercise**
>
> We need to combine `df` and `fg_df` using the pandas `merge()` function. As you prepare to do so, keep the following considerations front of mind:
> 1. Which type of join should you use to capture all of the information in both datasets? (**Hint:** Look at the `head` and `tail` of the resulting `DataFrame` for clues.)
> 2. In order to put the `FoodGroup` column immediately after the `NDB_No` column, in what order should you enter the two `DataFrame`s into the `merge()` function? (You might need to experiment a couple of times to get the desired order.)
>
> Perform the command to join the `df` and `fg_df` in the code cell below.

> **Exercise solution**
>
> The correct code to use for this is `combined_df = pd.merge(fg_df, df, how='outer', on='NDB_No')`.

In [ ]:
combined_df.head()

In [ ]:
combined_df.tail()

Now save the merged `DataFrame` using the `to_csv()` method.

In [ ]:
combined_df.to_csv('Data/USDA-nndb-merged.csv', 
                   sep=',', 
                   encoding='latin_1', 
                   index=False)

> **Takeaway:** Because the most interesting insights come from joining different datasets, the pandas `merge()` function is at the heart of most data science projects.